html_text/html_text.py

```diff
@@ -44,30 +53,108 @@ def parse_html(html):
-                                             not _has_punct_before(prev))):
-                    yield ' '
-                yield text
-                prev = text
-
-        return _whitespace.sub(' ', ''.join(fragments()).strip())
 
+    def add_space(text, prev):
+        if prev is None:
+            return False
+        if prev == '\n' or prev == '\n\n':
+            return False
+        if not _has_trailing_whitespace(prev):
+            if _has_punct_after(text) or _has_open_bracket_before(prev):
+                return False
+        return True
+
+    def add_newline(tag, prev):
+        if prev is None or prev == '\n\n':
+            return '', '\n\n'
+        if tag in double_newline_tags:
+            if prev == '\n':
+                return '\n', '\n\n'
+            return '\n\n', '\n\n'
+        if tag in newline_tags:
+            if prev == '\n':
+                return '', prev
+            return '\n', '\n'
+        return '', prev
+
+    def traverse_text_fragments(tree, prev, depth):
+        space = ' '
+        newline = ''
+        text = ''
+        if guess_page_layout:
+            newline, prev[0] = add_newline(tree.tag, prev[0])
+        if tree.text:
+            text = _whitespace.sub(' ', tree.text.strip())
+            if text and guess_punct_space and not add_space(text, prev[0]):
+                space = ''
+        if text:
+            yield [newline, space, text]
+            prev[0] = tree.text
+            space = ' '
+            newline = ''
+        elif newline:
+            yield [newline]
+            newline = ''
+
+        for child in tree:
+            for t in traverse_text_fragments(child, prev, depth + 1):
+                yield t
+
+        if guess_page_layout:
+            newline, prev[0] = add_newline(tree.tag, prev[0])
+
+        tail = ''
+        if tree.tail and depth != 0:
+            tail = _whitespace.sub(' ', tree.tail.strip())
+            if tail:
+                if guess_punct_space and not add_space(tail, prev[0]):
+                    space = ''
+        if tail:
+            yield [newline, space, tail]
+            prev[0] = tree.tail
+        elif newline:
+            yield [newline]
+
+    text = []
+    for fragment in traverse_text_fragments(tree, [None], 0):
+        text.extend(fragment)
+    return ''.join(text).strip()
+
+
+def selector_to_text(sel, guess_punct_space=True, guess_page_layout=False):
+    """ Convert a cleaned selector to text.
+    See html_text.extract_text docstring for description of the approach
+    and options.
+    """
+    if isinstance(sel, list):
+        # if selecting a specific xpath
+        text = []
+        for t in sel:
+            extracted = _html_to_text(
+                t.root,
+                guess_punct_space=guess_punct_space,
+                guess_page_layout=guess_page_layout)
+            if extracted:
+                text.append(extracted)
+        return ' '.join(text)
```

Comment:
This is to have it work as the previous implementation however I think it would make more sense to have it return a list of the text extracted by each selector. This way the user can decide whether and how to join it. Maybe they need all text as separate entities and that's why they want to select specific elements.

Reply:
I agree it is not clear which behavior is better; let's keep the current one, but have it in mind.